File: flutter_app/lib/ear/sight_singing_repository.dart
import 'dart:math';
import 'package:uuid/uuid.dart';

import 'sight_singing_models.dart';

/// 视唱会话数据源。
abstract class SightSingingRepository {
  Future<SightSingingSessionStart> startSession({
    required String pitchRange,
    required bool includeAccidental,
    required int questionCount,
  });

  Future<void> submitAnswer({
    required String sessionId,
    required String questionId,
    required List<String> answers,
    required double avgCentsAbs,
    required int stableHitMs,
    required int durationMs,
  });

  Future<SightSingingQuestion?> nextQuestion(String sessionId);
  Future<SightSingingResult> fetchResult(String sessionId);
}

/// 本地离线版视唱仓库：在内存中维护单次训练会话与结果。
class LocalSightSingingRepository implements SightSingingRepository {
  LocalSightSingingRepository({Uuid? uuid, Random? random})
      : _uuid = uuid ?? const Uuid(),
        _random = random ?? Random();

  final Uuid _uuid;
  final Random _random;
  final Map<String, _LocalSession> _sessions = <String, _LocalSession>{};

  @override
  Future<SightSingingSessionStart> startSession({
    required String pitchRange,
    required bool includeAccidental,
    required int questionCount,
  }) async {
    final range = _noteRangeFor(pitchRange);
    final config = SightSingingConfig(
      minNote: range.minNote,
      maxNote: range.maxNote,
      questionCount: questionCount,
      includeAccidental: includeAccidental,
    );
    final sessionId = _uuid.v4();
    final notes = _buildQuestionNotes(
      config: config,
      includeAccidental: includeAccidental,
      questionCount: questionCount,
    );
    final session = _LocalSession(config: config, questionNotes: notes);
    _sessions[sessionId] = session;
    return SightSingingSessionStart(
      sessionId: sessionId,
      config: config,
      question: _questionFor(session, 0),
    );
  }

  @override
  Future<void> submitAnswer({
    required String sessionId,
    required String questionId,
    required List<String> answers,
    required double avgCentsAbs,
    required int stableHitMs,
    required int durationMs,
  }) async {
    final session = _sessions[sessionId];
    if (session == null) {
      throw SightSingingApiException('训练会话不存在，请重新开始');
    }
    if (questionId != 'q-${session.currentIndex + 1}') {
      throw SightSingingApiException('题目状态已变化，请重新开始本题');
    }
    final isCorrect = avgCentsAbs <= 30 && stableHitMs >= 700;
    session.answered += 1;
    if (isCorrect) {
      session.correct += 1;
    }
  }

  @override
  Future<SightSingingQuestion?> nextQuestion(String sessionId) async {
    final session = _sessions[sessionId];
    if (session == null) {
      throw SightSingingApiException('训练会话不存在，请重新开始');
    }
    session.currentIndex += 1;
    if (session.currentIndex >= session.questionNotes.length) {
      return null;
    }
    return _questionFor(session, session.currentIndex);
  }

  @override
  Future<SightSingingResult> fetchResult(String sessionId) async {
    final session = _sessions.remove(sessionId);
    if (session == null) {
      throw SightSingingApiException('训练会话不存在，请重新开始');
    }
    final total = session.questionNotes.length;
    final answered = session.answered.clamp(0, total);
    final correct = session.correct.clamp(0, answered);
    final accuracy = answered == 0 ? 0.0 : correct / answered;
    return SightSingingResult(
      answered: answered,
      correct: correct,
      total: total,
      accuracy: accuracy,
    );
  }

  SightSingingQuestion _questionFor(_LocalSession session, int index) {
    return SightSingingQuestion(
      id: 'q-${index + 1}',
      index: index + 1,
      totalQuestions: session.questionNotes.length,
      targetNotes: <String>[session.questionNotes[index]],
    );
  }

  List<String> _buildQuestionNotes({
    required SightSingingConfig config,
    required bool includeAccidental,
    required int questionCount,
  }) {
    final candidates = _candidateNotes(
      minNote: config.minNote,
      maxNote: config.maxNote,
      includeAccidental: includeAccidental,
    );
    if (candidates.isEmpty) {
      return List<String>.filled(questionCount, 'C4');
    }
    return List<String>.generate(
      questionCount,
      (_) => candidates[_random.nextInt(candidates.length)],
    );
  }

  List<String> _candidateNotes({
    required String minNote,
    required String maxNote,
    required bool includeAccidental,
  }) {
    final min = _noteNameToMidi(minNote);
    final max = _noteNameToMidi(maxNote);
    final start = min <= max ? min : max;
    final end = min <= max ? max : min;
    final out = <String>[];
    for (var midi = start; midi <= end; midi++) {
      final note = _midiToNoteName(midi);
      if (!includeAccidental && note.contains('#')) {
        continue;
      }
      out.add(note);
    }
    return out;
  }

  _LocalRange _noteRangeFor(String range) {
    switch (range) {
      case 'low':
        return const _LocalRange(minNote: 'C3', maxNote: 'B3');
      case 'wide':
        return const _LocalRange(minNote: 'C3', maxNote: 'B4');
      case 'mid':
      default:
        return const _LocalRange(minNote: 'C4', maxNote: 'B4');
    }
  }
}

class SightSingingApiException implements Exception {
  SightSingingApiException(this.message);
  final String message;

  @override
  String toString() => message;
}

class _LocalSession {
  _LocalSession({required this.config, required this.questionNotes});

  final SightSingingConfig config;
  final List<String> questionNotes;
  int currentIndex = 0;
  int answered = 0;
  int correct = 0;
}

class _LocalRange {
  const _LocalRange({required this.minNote, required this.maxNote});

  final String minNote;
  final String maxNote;
}

int _noteNameToMidi(String note) {
  final n = note.trim().toUpperCase();
  final m = RegExp(r'^([A-G])(#?)(\d)$').firstMatch(n);
  if (m == null) return 60;
  final name = '${m.group(1)}${m.group(2) ?? ''}';
  final octave = int.tryParse(m.group(3) ?? '4') ?? 4;
  const names = <String>[
    'C',
    'C#',
    'D',
    'D#',
    'E',
    'F',
    'F#',
    'G',
    'G#',
    'A',
    'A#',
    'B',
  ];
  final idx = names.indexOf(name);
  return (octave + 1) * 12 + (idx < 0 ? 0 : idx);
}

String _midiToNoteName(int midi) {
  const names = <String>[
    'C',
    'C#',
    'D',
    'D#',
    'E',
    'F',
    'F#',
    'G',
    'G#',
    'A',
    'A#',
    'B',
  ];
  final note = names[midi % 12];
  final octave = midi ~/ 12 - 1;
  return '$note$octave';
}

File: swift_app/Sources/Features/Ear/SightSingingRepository.swift
import Foundation
import Core

public protocol SightSingingRepository: Sendable {
    func startSession(
        pitchRange: String,
        includeAccidental: Bool,
        questionCount: Int,
        exerciseKind: SightSingingExerciseKind
    ) async throws -> SightSingingSessionStart

    func submitAnswer(
        sessionId: String,
        questionId: String,
        answers: [String],
        avgCentsAbs: Double,
        stableHitMs: Int,
        durationMs: Int
    ) async throws

    func nextQuestion(sessionId: String) async throws -> SightSingingQuestion?
    func fetchResult(sessionId: String) async throws -> SightSingingResult
}

public enum SightSingingRepositoryError: Error, LocalizedError {
    case sessionNotFound
    case questionOutOfDate

    public var errorDescription: String? {
        switch self {
        case .sessionNotFound:
            return "训练会话不存在，请重新开始"
        case .questionOutOfDate:
            return "题目状态已变化，请重新开始本题"
        }
    }
}

public actor LocalSightSingingRepository: SightSingingRepository {
    private final class LocalSession: @unchecked Sendable {
        let config: SightSingingConfig
        let questionTargets: [[String]]
        var currentIndex = 0
        var answered = 0
        var correct = 0

        init(config: SightSingingConfig, questionTargets: [[String]]) {
            self.config = config
            self.questionTargets = questionTargets
        }
    }

    private var sessions: [String: LocalSession] = [:]
    private var rng = SystemRandomNumberGenerator()

    public init() {}

    public func startSession(
        pitchRange: String,
        includeAccidental: Bool,
        questionCount: Int,
        exerciseKind: SightSingingExerciseKind
    ) async throws -> SightSingingSessionStart {
        let range = noteRange(for: pitchRange)
        let config = SightSingingConfig(
            minNote: range.minNote,
            maxNote: range.maxNote,
            questionCount: questionCount,
            includeAccidental: includeAccidental,
            exerciseKind: exerciseKind
        )
        let targets = buildQuestionTargets(config: config)
        let sessionId = UUID().uuidString.lowercased()
        let session = LocalSession(config: config, questionTargets: targets)
        sessions[sessionId] = session
        return SightSingingSessionStart(
            sessionId: sessionId,
            config: config,
            question: question(for: session, index: 0)
        )
    }

    public func submitAnswer(
        sessionId: String,
        questionId: String,
        answers: [String],
        avgCentsAbs: Double,
        stableHitMs: Int,
        durationMs: Int
    ) async throws {
        guard let session = sessions[sessionId] else {
            throw SightSingingRepositoryError.sessionNotFound
        }
        if questionId != "q-\(session.currentIndex + 1)" {
            throw SightSingingRepositoryError.questionOutOfDate
        }
        let isCorrect = avgCentsAbs <= 30 && stableHitMs >= 700
        session.answered += 1
        if isCorrect {
            session.correct += 1
        }
    }

    public func nextQuestion(sessionId: String) async throws -> SightSingingQuestion? {
        guard let session = sessions[sessionId] else {
            throw SightSingingRepositoryError.sessionNotFound
        }
        session.currentIndex += 1
        guard session.currentIndex < session.questionTargets.count else {
            return nil
        }
        return question(for: session, index: session.currentIndex)
    }

    public func fetchResult(sessionId: String) async throws -> SightSingingResult {
        guard let session = sessions.removeValue(forKey: sessionId) else {
            throw SightSingingRepositoryError.sessionNotFound
        }
        let total = session.questionTargets.count
        let answered = min(total, max(0, session.answered))
        let correct = min(answered, max(0, session.correct))
        let accuracy = answered == 0 ? 0 : Double(correct) / Double(answered)
        return SightSingingResult(answered: answered, correct: correct, total: total, accuracy: accuracy)
    }

    private func question(for session: LocalSession, index: Int) -> SightSingingQuestion {
        SightSingingQuestion(
            id: "q-\(index + 1)",
            index: index + 1,
            totalQuestions: session.questionTargets.count,
            targetNotes: session.questionTargets[index]
        )
    }

    private func buildQuestionTargets(config: SightSingingConfig) -> [[String]] {
        switch config.exerciseKind {
        case .singleNoteMimic:
            let singles = buildSingleQuestionNotes(config: config)
            return singles.map { [$0] }
        case .intervalMimic:
            return buildIntervalQuestionPairs(config: config)
        }
    }

    private func buildSingleQuestionNotes(config: SightSingingConfig) -> [String] {
        let minMidi = noteNameToMidi(config.minNote)
        let maxMidi = noteNameToMidi(config.maxNote)
        let lo = min(minMidi, maxMidi)
        let hi = max(minMidi, maxMidi)
        var candidates: [String] = []
        for midi in lo...hi {
            let note = midiToNoteName(midi)
            if !config.includeAccidental && note.contains("#") {
                continue
            }
            candidates.append(note)
        }
        guard !candidates.isEmpty else {
            return Array(repeating: "C4", count: config.questionCount)
        }
        return (0..<config.questionCount).map { _ in candidates.randomElement(using: &rng) ?? "C4" }
    }

    private func buildIntervalQuestionPairs(config: SightSingingConfig) -> [[String]] {
        let minMidi = noteNameToMidi(config.minNote)
        let maxMidi = noteNameToMidi(config.maxNote)
        let lo = min(minMidi, maxMidi)
        let hi = max(minMidi, maxMidi)
        var candidates: [String] = []
        for midi in lo...hi {
            let note = midiToNoteName(midi)
            if !config.includeAccidental && note.contains("#") {
                continue
            }
            candidates.append(note)
        }
        guard candidates.count >= 2 else {
            return Array(repeating: ["C4", "D4"], count: config.questionCount)
        }

        return (0..<config.questionCount).map { _ in
            // 随机上行两音（允许同音不同八度，但不允许完全同 MIDI）。
            for _ in 0..<48 {
                let aName = candidates.randomElement(using: &rng) ?? "C4"
                let bName = candidates.randomElement(using: &rng) ?? "D4"
                let a = noteNameToMidi(aName)
                let b = noteNameToMidi(bName)
                if a == b { continue }
                let lowMidi = min(a, b)
                let highMidi = max(a, b)
                return [midiToNoteName(lowMidi), midiToNoteName(highMidi)]
            }
            // 兜底：选一个相邻半音对，保证可判定。
            let base = noteNameToMidi(candidates.randomElement(using: &rng) ?? "C4")
            let low = max(lo, min(base, hi - 1))
            let high = min(hi, low + 1)
            return [midiToNoteName(low), midiToNoteName(high)]
        }
    }

    private func noteRange(for pitchRange: String) -> (minNote: String, maxNote: String) {
        switch pitchRange {
        case "low":
            return ("C3", "B3")
        case "wide":
            return ("C3", "B4")
        default:
            return ("C4", "B4")
        }
    }

    private func noteNameToMidi(_ note: String) -> Int {
        let upper = note.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        let regex = try? NSRegularExpression(pattern: "^([A-G])(#?)(\\d)$")
        let range = NSRange(location: 0, length: upper.utf16.count)
        guard
            let match = regex?.firstMatch(in: upper, range: range),
            match.numberOfRanges == 4,
            let n1 = Range(match.range(at: 1), in: upper),
            let n2 = Range(match.range(at: 2), in: upper),
            let n3 = Range(match.range(at: 3), in: upper)
        else {
            return 60
        }
        let name = String(upper[n1]) + String(upper[n2])
        let octave = Int(upper[n3]) ?? 4
        let names = PitchMath.noteNames
        let idx = names.firstIndex(of: name) ?? 0
        return (octave + 1) * 12 + idx
    }

    private func midiToNoteName(_ midi: Int) -> String {
        let note = PitchMath.noteNames[((midi % 12) + 12) % 12]
        let octave = midi / 12 - 1
        return "\(note)\(octave)"
    }
}
